### ISSUE TITLE
OCSP response contains "Bad time value" in "Produced At:"

### ISSUE THREAD

**Issue by kinichiro** [libressl-portable/portable#201]
Hi,

Now, I'm testing OCSP with `openssl ocsp` command,
and I found that OCSP response seems to be broken.

I tested with this script, and test results are also posted here.
https://gist.github.com/kinichiro/882924e18665b20495f079b14b63a828
This script generates OCSP response, and verify it by `-respin` and `-resp_text` option.

I've got "Bad time value" with LibreSSL 2.4.1 on CentOS7(x86_64).
<pre>
...
[TEST] ocsp ... verify OCSP response
OCSP Response Data:
    OCSP Response Status: successful (0x0)
    Response Type: Basic OCSP Response
    Version: 1 (0x0)
    Responder Id: C = JP, ST = Tokyo, O = TEST_DUMMY_COMPANY, CN = testOCSP.test_dummy.com
Response verify OK
    Produced At: Bad time value:-) success.
</pre>

OpenSSL 1.0.1e-fips works fine.
<pre>
...
[TEST] ocsp ... verify OCSP response
OCSP Response Data:
    OCSP Response Status: successful (0x0)
    Response Type: Basic OCSP Response
    Version: 1 (0x0)
    Responder Id: C = JP, ST = Tokyo, O = TEST_DUMMY_COMPANY, CN = testOCSP.test_dummy.com
    Produced At: Jun 25 04:18:49 2016 GMT
    Responses:
    Certificate ID:
      Hash Algorithm: sha1
      Issuer Name Hash: B489BF1512CC8520061B55B53829E74BE05B4574
...
</pre>

Thanks.

**Comment by kinichiro** [comment]
I checked response DER file with 
`openssl asn1parse -in ocsptest/user1/ocsp_res.der -inform der -strparse 26` command.
Then I noticed that 1 GENERALIZEDTIME field is set with UTCTIME format.

Here is the result by LibreSSL 2.4.1.
<pre>
   74:d=6  hl=2 l=   3 prim: OBJECT            :commonName
   79:d=6  hl=2 l=  23 prim: UTF8STRING        :testOCSP.test_dummy.com
  104:d=2  hl=2 l=  13 prim: GENERALIZEDTIME   :160625041135Z
</pre>

Here is the result by OpenSSL 1.0.1e.
<pre>
   74:d=6  hl=2 l=   3 prim: OBJECT            :commonName
   79:d=6  hl=2 l=  23 prim: T61STRING         :testOCSP.test_dummy.com
  104:d=2  hl=2 l=  15 prim: GENERALIZEDTIME   :20160625041849Z
</pre>

**Comment by kinichiro** [comment]
According to RFC, `producedAt` should be the type of `GeneralizedTime`.
https://tools.ietf.org/html/rfc6960

`X509_gmtime_adj()` generates UTCTIME if year < 2050 for RFC5280.
I think `ASN1_GENERALIZEDTIME_set()` can be used instead of `X509_gmtime_adj()`, like below.

```diff
diff --git a/src/lib/libssl/src/crypto/ocsp/ocsp_srv.c b/src/lib/libssl/src/crypto/ocsp/ocsp_srv.c
index 5629396..86e8c94 100644
--- a/src/lib/libssl/src/crypto/ocsp/ocsp_srv.c
+++ b/src/lib/libssl/src/crypto/ocsp/ocsp_srv.c
@@ -260,7 +260,7 @@ OCSP_basic_sign(OCSP_BASICRESP *brsp, X509 *signer, EVP_PKEY *key,
        }

        if (!(flags & OCSP_NOTIME) &&
-           !X509_gmtime_adj(brsp->tbsResponseData->producedAt, 0))
+           !ASN1_GENERALIZEDTIME_set(brsp->tbsResponseData->producedAt, time(NULL)))
                goto err;

        /* Right now, I think that not doing double hashing is the right
```

**Comment by kinichiro** [comment]
Reference http://permalink.gmane.org/gmane.comp.lang.ruby.cvs/58647

**Comment by bob-beck** [comment]
Yes, kinichiro - This looks correct to me.

**Comment by kinichiro** [comment]
Fixed by https://github.com/libressl-portable/openbsd/commit/89887f8c015d3d2dd2b6300bf41b7bb3d30d18a8
Thanks.

**Comment by kinichiro** [comment]
I verified 2 OCSP fixes (#201 and #202) were valid for ruby, too. Here is the report for that.

At first, build ruby with LibreSSL.
<pre>
# # build and install LibreSSL current
#
# cd work/libressl-2.4.x/
# ./configure --prefix=/usr/local/libressl-2.4.x
# make check
# make install
#
# # build and test ruby 2.4.0dev
#
# git clone https://github.com/ruby/ruby/ work/ruby
# cd work/ruby/
# autoconf
# ./configure --with-opt-dir=/usr/local/libressl-2.4.x
# make
#
</pre>

Next, run the openssl test of ruby.

with LibreSSL-2.3.6, 2 tests were SKIPPED, like below.
<pre>
# make test-all TESTS=openssl TESTOPTS="--show-skip"
...
  1) Skipped:
OpenSSL::TestOCSP#test_basic_response_der [/work/github/ruby/test/openssl/test_ocsp.rb:145]:
OCSP_basic_sign() is broken

  2) Skipped:
OpenSSL::TestOCSP#test_single_response_check_validity [/work/github/ruby/test/openssl/test_ocsp.rb:216]:
OCSP_check_validity() is broken

Finished tests in 18.046764s, 21.7768 tests/s, 395.6942 assertions/s.
393 tests, 7141 assertions, 0 failures, 0 errors, 2 skips

ruby -v: ruby 2.4.0dev (2016-06-25 trunk 55502) [x86_64-linux]
</pre>

with LibreSSL current, 0 skipped test, like this.
<pre>
# make test-all TESTS=openssl TESTOPTS="--show-skip"
...
Finished tests in 18.315690s, 21.4570 tests/s, 390.3757 assertions/s.
393 tests, 7150 assertions, 0 failures, 0 errors, 0 skips

ruby -v: ruby 2.4.0dev (2016-06-25 trunk 55502) [x86_64-linux]
# 
</pre>

Best Regards,
